src/bootsampling.jl

```diff
@@ -337,55 +335,50 @@ function bootstrap(statistic::Function, data, model::SimpleModel, sampling::Boot
     return ParametricBootstrapSample(t0, t1, statistic, data, model, sampling)
 end
 
-
-"""
-bootstrap(statistic, data, model, formula, sampling)
-"""
-function bootstrap(statistic::Function, data::AbstractDataFrame, model::FormulaModel, sampling::ResidualSampling)
-    class = model.class
-    formula = model.formula
-    args = model.args
-    kwargs = model.kwargs
-    yy = lhs(formula)
-    y0 = data[yy]
-    f0 = fit(class, formula, data, args...; kwargs...)
-    t0 = tx(statistic(f0))
-    r0 = predict(f0) - y0
-    m = nrun(sampling)
-    t1 = zeros_tuple(t0, m)
-    r1 = copy(r0)
-    data1 = deepcopy(data)
-    for i in 1:m
-        sample!(r0, r1)
-        data1[:,yy] = y0 + r1
-        f1 = fit(class, formula, data1, args...; kwargs...)
-        for (j, t) in enumerate(tx(statistic(f1)))
-            t1[j][i] = t
-        end
-    end
-    return ParametricBootstrapSample(t0, t1, statistic, data, model, sampling)
+struct ResidualTerm{S,T,R} <: StatsModels.AbstractTerm
+    sampling::S
+    t::T
+    r0::R
+    r1::R
+    ResidualTerm(s::S, t::T, r::R) where {S,T,R} = new{S,T,R}(s, t, r, copy(r))
 end
 
 
+StatsModels.modelcols(t::ResidualTerm{S}, data) where {S} =
+    throw(ArgumentError("ResidualTerm: don't know how to generate model columns for $S"))
+StatsModels.modelcols(t::ResidualTerm{ResidualSampling}, data) =
+    modelcols(t.t, data) + sample!(t.r0, t.r1)
+StatsModels.modelcols(t::ResidualTerm{WildSampling}, data) =
+    modelcols(t.t, data) + t.sampling.noise(t.r0)
+
+# we only ever create one of these terms after the model has been fit once so we don't
+# need to create a new schema for it every time...
+StatsModels.needs_schema(t::ResidualTerm) = false
+StatsModels.termvars(t::ResidualTerm) = StatsModels.termvars(t.t)
+
 """
+bootstrap(statistic, data, model, formula, sampling)
 bootstrap(statistic, data, model, formula, Wildsampling(nrun, noise))
 """
-function bootstrap(statistic::Function, data::AbstractDataFrame, model::FormulaModel, sampling::WildSampling)
```

Comment:
Can we constrain `sampling <: ParametricSampling` here?

Reply:
You'd need to make Wild/ResidualSampling descendants of that though right?  RIght now they're both `<:BootstrapSampling` which is the supertype of Parametric (unless I'm missing something)

Reply:
Actually as far as I can tell nothing is `<:[Non]ParametricSampling` at the moment...

Reply:
Yeah, you are right. I wonder why `[Non]ParametricSampling` aren't used - maybe I forgot the use them in the signatures or I removed them partially at a later stage. I'll need to investigate this.
Can we then make it `<:BootstrapSampling`?

Reply:
Got it, just pushed that change